libcalico-go/lib/backend/k8s/resources/resources.go

```diff
@@ -293,7 +305,9 @@ func ConvertK8sResourceToCalicoResource(res Resource) error {
 
 	// Manually write in the data not stored in the annotations: Name, Namespace, ResourceVersion,
 	// so that they do not get overwritten.
-	meta.Name = rom.GetName()
+	if meta.Name == "" {
```

Comment:
```suggestion
	if meta.Name == "" {
	  // We store the original v3 Name in our annotation when writing NetworkPolicy and GNP objects. If that's present, use it.
	  // Otherwise, fill in the name from the underlying CRD.
```